### PR TITLE
[FIX] Passing HttpException from Exception handler

### PR DESCRIPTION
**Pull request by geoperez**
From SO question at https://stackoverflow.com/questions/58718698/how-to-setup-an-exception-handler-in-embedio-web-api

### CODE REVIEW

File: src/EmbedIO/ExceptionHandler.cs [hunk 1]
```diff
@@ -137,6 +137,10 @@ internal static async Task Handle(string logSource, IHttpContext context, Except
                 await handler(context, exception)
                     .ConfigureAwait(false);
             }
+            catch (HttpException)
```
Comment (rdeago) [inline src/EmbedIO/ExceptionHandler.cs]: `catch (Exception exception) when (exception is IHttpException)`

**Comment by geoperez** [comment]
@rdeado, I saw your response in SO, did you test that code? Because I first tried throwing `HttpException` from `HandleUnhandledException` callback, but this always returns 404.

**Comment by geoperez** [comment]
One issue is that you can only set a proper handler for the `HttpException` in the module's parent. Like this:

```
            var server = new WebServer(o => o
                    .WithUrlPrefix(url)
                    .WithMode(HttpListenerMode.EmbedIO)).WithWebApi("/api", o =>
            {
                o.WithController<TestController>();
                
                o.OnUnhandledException = (ctx, ex) => throw HttpException.BadRequest();
            });
            
            server.OnHttpException = async (ctx, ex) => await ctx.SendDataAsync(ex);
```

The following code, the  proper way, is not working:

```

```
            var server = new WebServer(o => o
                    .WithUrlPrefix(url)
                    .WithMode(HttpListenerMode.EmbedIO)).WithWebApi("/api", o =>
            {
                o.WithController<TestController>();
                
                o.OnHttpException = async (ctx, ex) => await ctx.SendDataAsync(ex);
                o.OnUnhandledException = (ctx, ex) => throw HttpException.BadRequest();
            });
```

**Comment by rdeago** [comment]
I have to admit that the code I posted on SO was only submitted to the looks-legit-may even-compile™️ test. Anyway, the behavior you describe looks more like a bug to me than a need for an additional `catch` clause.

**Comment by geoperez** [comment]
A bug because it is not passing to the `OnHttpException` handler?

**Comment by geoperez** [comment]
@rdeago I changed the signature of `ExceptionHandler`, what do you think?

**Comment by rdeago** [comment]
About being only able to handle the resulting HTTP exception at an outer level: you are right, but think about it: next time someone wants to throw an exception from an HTTP exception handler, what do we do inside the `catch` clause you just added? Catch exceptions and pass them to the exception handler, which in turn can throw another HTTP exception?

Either we loop, handling exceptions and catching exceptions thrown by handlers until no further exception is thrown, or we handle exceptions thrown by handlers at an outer level; after all, structured exception handling got its name for a reason.

Should the need arise, in some particular setup, to throw HTTP exceptions from an exception handler and have them handled differently from HTTP exceptions generated in other modules, just wrap the module in a `ModuleGroup`, thus introducing a further handling layer.

File: src/EmbedIO/ExceptionHandler.cs [hunk 2]
```diff
@@ -120,7 +121,7 @@ public static Task HtmlResponse(IHttpContext context, Exception exception)
                     }
                 });
 
-        internal static async Task Handle(string logSource, IHttpContext context, Exception exception, ExceptionHandlerCallback? handler)
+        internal static async Task Handle(string logSource, IHttpContext context, Exception exception, ExceptionHandlerCallback? handler, HttpExceptionHandlerCallback? httpHandler)
```
Comment (rdeago) [inline src/EmbedIO/ExceptionHandler.cs]: What if I also want to handle exceptions thrown by the HTTP exception handler? One more `catch` block, one more handler... And what about HTTP exceptions thrown by _that_ handler? And so on, and so forth.

**Comment by geoperez** [comment]
I understand what are you trying to explain here. I know a `HttpExceptionHandler` can throw an unexpected exception, but in this scenario, you are in control of the handler.

In the user story of this PR, you might not have control of what are you connecting to your WebAPI, and it's really hard to translate from a common exception to one using `IHttpException`. The only solution it's having a try/catch in all your endpoints. Kinda messy. 

I may recommend keep this patch, and open the discussion if we need a more complex exception handler.

**Comment by rdeago** [comment]
My browser keeps crashing. I'm moving to Slack to try and continue the conversation.

Can you defer merging for now, so we can see if there's a cleaner way to make everyone happy?

File: src/EmbedIO/ExceptionHandler.cs (hunk 2 above)
Comment (rdeago) [reply src/EmbedIO/ExceptionHandler.cs]: OK we have clarified that on Slack. 👍

File: test/EmbedIO.Tests/Issues/Issue_PassHttpException.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+{
+    [TestFixture]
+    public class Issue_PassHttpException : EndToEndFixtureBase
+    {
+        protected override void OnSetUp()
+        {
+            Server.WithWebApi("/api", o =>
+                {
+                    o.WithController<TestController>();
+                    o.OnUnhandledException = (ctx, ex) => throw HttpException.BadRequest();
```
Comment (rdeago) [inline test/EmbedIO.Tests/Issues/Issue_PassHttpException.cs]: Maybe, just to be sure, use some really weird status code such as `417 I'm a teapot` or something.

Comment (geoperez) [reply test/EmbedIO.Tests/Issues/Issue_PassHttpException.cs]: Usually, this will return a 404.

**Pull request merged**